src/runtime/index.ts

```diff
@@ -1,7 +1,123 @@
-import platform from "./platform";
+import platform, { Platform } from "./platform";
+import { EventEmitter } from "node:events";
+import { Transport } from "./transport";
+import { MessageEvent } from "../proto/server/v1/realtime";
 
-export class Realtime {
+type SubscribeCallback = (string) => void;
+
+export interface RealTimeConfig {
+  /*
+   * An id used to identify this client when using the presence feature.
+   * If the client id is not set, an error will be thrown when using any presence features
+   */
+  clientId?: string;
+  platform: Platform;
+}
+
+export class RealTime {
+  private _config: RealTimeConfig;
+  private channelManager: ChannelManager;
+  private transport: Transport;
+  constructor(config: RealTimeConfig) {
+    this.transport = new Transport({
+      WebSocket: config.platform.WebSocket,
+      heartbeatTimeout: 1000,
+    });
+    this.channelManager = new ChannelManager(this.transport);
+
+    this._config = config;
+  }
+
+  async connect() {
+    await this.transport.isConnected();
+  }
+
+  getChannel(name: string): Channel {
+    return this.channelManager.getOrCreate(name);
+  }
+
+  close() {
+    this.channelManager.close();
+    this.transport.close();
+  }
+
+  socketId() {
+    return this.transport.socketId();
+  }
+}
+
+class Presence extends EventEmitter {
   constructor() {
-    console.log("constructor", platform.runtime, platform.WebSocket);
+    super();
+  }
+}
+
+export class Channel extends EventEmitter {
+  private name: string;
+  private hasAttached: boolean;
+  private transport: Transport;
+
+  constructor(name: string, transport: Transport) {
+    super();
+    this.hasAttached = false;
+    this.name = name;
+    this.transport = transport;
+  }
+
+  subscribe(msgName: string, cb: SubscribeCallback) {
+    if (!this.hasAttached) {
+      this.attach();
+    }
+
+    this.on(msgName, cb);
+  }
+
+  unsubscribe(msgName: string, cb: SubscribeCallback) {
+    this.off(msgName, cb);
+  }
+
+  unsubscribeAll(msgName: string) {
+    this.removeAllListeners(msgName);
+  }
+
+  attach() {
+    this.transport.listen(this.name, (msg: MessageEvent) => this.notify(msg));
+    this.hasAttached = true;
+  }
+
+  detach() {
+    this.hasAttached = false;
+  }
+
+  async publish(msgName: string, data: string) {
```

Comment:
so would data be a string or any? as the user can pass random bytes as well?

Reply:
I think string is best for now.

Reply:
So the only caveat is if the user needs to store some obj, this will then push the logic on the caller side to do the encoding first?

Reply:
I've changed this so that the data is encoded via msgpack to be a byte array.